text/projects/Maciej_Nems/Simple_String_Matching_k_Mistakes.tex

```diff
@@ -0,0 +1,126 @@
+\par
+Teraz zdefiniujmy funkcję $MISMATCH(j, m, k, lcp)$, która dla
+\begin{itemize}
+    \item indeksu $j$ w $W@\tilde{T}$
+    \item $m,k$ z problemu SMK
+    \item $lcp$ będącego strukturą odpowiadającą na problem longest common prefix dla drzewa sufiksowego $W@\tilde{T}$ w czasie stałym. $lcp.query(i,j)$ zwraca długość najdłuższego wspólnego prefiksu dla sufiksów $W@\tilde{T}$ zaczynających się na pozycjach $i$ oraz $j$
+\end{itemize}
+zwraca \textbf{true} wtw, gdy odległość hamminga dla podsłów zaczynających się na pozycjach $1$ oraz $j$ w $W@\tilde{T}$ wynosi $\leq k$. Mając taki algorytm, dla każdego znalezionego poddrzewa $u$, po znalezieniu liścia $x$, sprawdzamy, czy $x$ jest oznaczony, oraz czy $MISMATCH(j_x,m,k,lcp)$ zwraca \textbf{true}. Jeśli tak, to znaczy, że wszystkie liście w poddrzewie $u$ (oprócz liścia odpowiadającego całemu słowu) są dopasowaniami $W$ z $k$ błędami, więc możemy zwrócić ich indeksy w $\tilde{T}$.
+\begin{algorithm}[H]
+\caption{MISMATCH}
+\begin{algorithmic}
+\INPUT{$j,m,k,lcp$}
+\STATE{$w=1$\ \ \ // represents current index in pattern part of $W@\tilde{T}$}
+\STATE{$t=j$\ \ \ // represents current index in text part of $W@\tilde{T}$}
+\STATE{$c=0$\ \ \ // counter for mismatches}
+\STATE{// until there are more than $k$ mistakes, or index in pattern is $>m$}
+\STATE{// we find longest common prefix and update indexes for pattern and text}
+\WHILE{$w\leq m$ \textbf{and} $c\leq k$}
+    \STATE{$q:= lcp.query(w,t)$}
+    \IF{$w+q\leq m$}
+        \STATE{// LCP ended before end of pattern, we should increment mistake counter}
+        \STATE{$c = c + 1$}
+    \ENDIF
+    \STATE{w = w + q + 1}
+    \STATE{t = t + q + 1}
+\ENDWHILE
+\RETURN{$c \leq k$}
+\end{algorithmic}
+\end{algorithm}
+\par
+\subsubsection*{Poprawność}
+Dlaczego oznaczamy $W@\tilde{T}$? Ponieważ jako jedyny sufiks zaczynający się w części $W$ może mieć prefiks długości $m$ wspólny z sufiksami zaczynającymi się w części $\tilde{T}$ (pozostałe mają $@$ na $m$ pierwszych znakach). Zauważmy, że dla każdego poddrzewa $u$, które rozpatrywaliśmy w algorytmie, albo wszystkie liście są oznaczone, albo wszystkie liście nie są oznaczone (mają ten sam prefiks długości $m$, więc dla wszystkich z nich permutation matching musiał znaleźć to samo). Przeglądając drzewo w kolejności preorder znajdziemy wszystkie chciane $u$ (odpowiada prefiksowi długosći $\geq m$). Dodatkowo, łatwo zauważyć, że funkcja MISMATCH poprawnie oblicza, czy odległość hamminga takiego wspólnego prefiksu jest $\leq k$. Jeśli więc dla każdego takiego $u$ sprawdzimy dowolny liść, czy został oznaczony i wartość funkcji MISMATCH, to znajdziemy wszystkie indeksy dopasowań z co najwyżej $k$ błędami.
```

Comment:
"Dlaczego oznaczamy $W@\tilde{T}$? Ponieważ jako jedyny sufiks zaczynający się w części $W$ może mieć prefiks długości $m$ wspólny z sufiksami zaczynającymi się w części $\tilde{T}$ (pozostałe mają $@$ na $m$ pierwszych znakach)." - that one I don't get too, please rephrase it

Reply:
What about:
Dlaczego oznaczamy $W@\tilde{T}$? Zauważmy, że to co robi ten algorytm, to szukanie oznaczonych liści, które mają prefiks długości $m$ z odpowiednią odległością Hamminga. Jako, że  $W@\tilde{T}$ ma prefiks długości $m$ z odległością Hamminga $0$ (część $W$), to w drzewie sufiksowym może wystąpić w tym samym poddrzewie, co faktyczne liście, które powinniśmy zliczyć. Ten prefiks nie jest jednak dopasowaniem, ponieważ nie leży w całości w $T$. Pozostałe prefiksy, które są "nielegalne" (nie leżą w całości w $T$) nigdy nie wylądują w tym samym poddrzewie, ponieważ mają znak $@$, który nie występuje w $T$.

Zauważmy, że dla każdego poddrzewa $u$, które rozpatrywaliśmy w algorytmie...

Reply:
OK, now it's better